# Hashing, Encryption, Blockchain & Bitcoin Mining with Python
https://www.youtube.com/watch?v=8ke8xn60MIc

### hash = A hash function is any function that can be used to map data of arbitrary size to data of fixed size. The values returned by a hash function are called hash values, hash codes, digests, or simply hashes.

#### first simple hash, maps any string to a three digit integer.  It uses ordinal numbers of one character string objects

In [1]:
ord('a')

97

In [2]:
chr(97)

'a'

## The implementation of te simplistic Hash Function ("average integer ordina number")

In [3]:
def hash_function(text):
    value = sum((ord(l) for l in text))/len(text)
    return '%03d' % value ##'%03d' Means 3 Digits!!

##### '%03d' will return values in 3 digits
##### for x in y worth looking into more

In [4]:
hash_function('Jonathan Tompkins')

'100'

#### Our function has a target space of 10 bits only

In [5]:
bin(999)

'0b1111100111'

In [6]:
2 ** 10

1024

#### Modern hash functions have a target space of 128 bits, i.e. 2^128 -1 possible values

In [7]:
2 ** 128

340282366920938463463374607431768211456

In [8]:
hex(2 ** 128)

'0x100000000000000000000000000000000'

In [9]:
len(hex(2 ** 128)) - 4

31

#### Or 256 bits

In [10]:
2 ** 256

115792089237316195423570985008687907853269984665640564039457584007913129639936

In [11]:
len(hex(2 ** 256)) - 4

63

#### Some even have 384 bits

In [12]:
len(hex(2 ** 384)) - 4

95

#### The universe is assumed to consist of 10^80 atoms

In [13]:
10 ** 80

100000000000000000000000000000000000000000000000000000000000000000000000000000000

In [14]:
len(hex(10 ** 80)) - 4  # close to 2 ** 256

65

In [15]:
hex(10 ** 80)

'0x35f9dea3e1f6bdfef70cdd17b25efa418ca63a22764cec100000000000000000000'

## We Require the following properties from a hash function
    Collision Resistance - highly unlikely for inputs to yield same output
    hiding - highly difficult to find input if you know output
    puzzle friendliness - If someone targets a certain output value and if parts of the input are randomly chosen then it is difficult to find another input value that hits exactly that target

## MD5 Hash Codes

Still used, not recommended for storing passwords, etc

In [16]:
import hashlib

In [17]:
md5_1 = hashlib.md5(b'jontomp').hexdigest()
md5_1

'7827ad9a7412cf9c2d38aee43393e480'

In [18]:
md5_2 = hashlib.md5(b'tompjon').hexdigest()
md5_2

'1ebc0892f040fd6b3ae6e9e6b26ce0f3'

#### Not Random, Deterministic!!! 

## Brute Force Hash (Password) Cracking

We define first a character set of lower case letters only

In [19]:
import string
charset = string.ascii_lowercase
charset

'abcdefghijklmnopqrstuvwxyz'

#### Hash Codes for all single characters in hashset

In [20]:
for c in charset:
    md5 = hashlib.md5(c.encode('ascii'))
    print(c, md5.hexdigest())

a 0cc175b9c0f1b6a831c399e269772661
b 92eb5ffee6ae2fec3ad71c777531578f
c 4a8a08f09d37b73795649038408b5f33
d 8277e0910d750195b448797616e091ad
e e1671797c52e15f763380b45e841ec32
f 8fa14cdd754f91cc6554c9e71929cce7
g b2f5ff47436671b6e533d8dc3614845d
h 2510c39011c5be704182423e3a695e91
i 865c0c0b4ab0e063e5caa3387c1a8741
j 363b122c528f54df4a0446b6bab05515
k 8ce4b16b22b58894aa86c421e8759df3
l 2db95e8e1a9267b7a1188556b2013b33
m 6f8f57715090da2632453988d9a1501b
n 7b8b965ad4bca0e41ab51de7b31363a1
o d95679752134a2d9eb61dbd7b91c4bcc
p 83878c91171338902e0fe0fb97a8c47a
q 7694f4a66316e53c8cdd9d9954bd611d
r 4b43b0aee35624cd95b910189b3dc231
s 03c7c0ace395d80182db07ae2c30f034
t e358efa489f58062f10dd7316b65649e
u 7b774effe4a349c6dd82ad4f4f21d34c
v 9e3669d19b675bd57058fd4664205d2a
w f1290186a5d0b1ceab27f4e77c0c5d68
x 9dd4e461268c8034f5c8564e155c67a6
y 415290769594460e2e485922904f345d
z fbade9e36a3f36d3d676c1b808451dd7


just need to chaeck 26 values if password is one lowercase number

#### Now doing brute force hash code cracking - 'knowing' that the relevant word has a maximum of 4 characters

In [21]:
import itertools as it
md5_3 = hashlib.md5(b'yves').hexdigest()

In [22]:
%%time
for i in range(1, 5):
    print('%d CHARACTERS USED NOW' % i)
    pm = it.product(charset, repeat = i)
    for comb in pm:
        comb = ''.join(comb)
        md5 = hashlib.md5(comb.encode('ascii')).hexdigest()
        if md5 == md5_3:
            print('SUCCESS')
            print(comb, md5)
            break

1 CHARACTERS USED NOW
2 CHARACTERS USED NOW
3 CHARACTERS USED NOW
4 CHARACTERS USED NOW
SUCCESS
yves afe3bd960b4c46a68580c4e564cca24e
Wall time: 869 ms


Let us now enlarge the character set to include digits as well

In [23]:
charset2 = string.ascii_lowercase + string.digits
charset2

'abcdefghijklmnopqrstuvwxyz0123456789'

Time to crack the second hash code increases tdue to greater password length and larger character set

In [24]:
from itertools import product
md5_4 = hashlib.md5(b'yves2').hexdigest()

In [25]:
import time
t0 = time.time()
z = 0
for i in range(1, 6):
    print('%d CHARACTERS USED NOW' % i)
    pm = it.product(charset2, repeat = i)
    for comb in pm:
        comb = ''.join(comb)
        md5 = hashlib.md5(comb.encode('ascii')).hexdigest()
        z += 1
        if md5 == md5_4:
            print('SUCCESS')
            print(comb, md5)
            break
sec = time.time() - t0
print('time in sec: %.1f' % sec)

1 CHARACTERS USED NOW
2 CHARACTERS USED NOW
3 CHARACTERS USED NOW
4 CHARACTERS USED NOW
5 CHARACTERS USED NOW
SUCCESS
yves2 664d839e06bada38ce04f7208896efdf
time in sec: 63.0


the algorithm has checked about 43 mm hashes efore being successful.  This represents a speed of about 450,000 hashes per second

In [26]:
z

43024025

In [27]:
z / sec

683107.4532942998

### Using dedicated password cracking tools

hashcat (http://hashcat.net) allows for a much faster and more intelligent/targeted approach

Rules
- password is 5 characters in length
- only lowercase letters or numbers

In [28]:
r1 = '''


Done in hashcat - cuts time down to 2.5 seconds

'''

### Make use of Human Traits ("Social Cracking")

Mask Attacks are somf of the most powerful tools in password cracking.  Human beings liek to use easy to remeber structures for passwords.
https://praetorian.com/blog/statistics-will-crack-your-password-mask-structure
- 50% of passwords rely on
- 13 password masks out of 
- 260,000+ masks analyzed in total

Example: lisa2008 = name of daughter born in 2008

**Last 4 digits are often years

Let us consider: a password is assumed to consist of uppercase letters, lowercase letters, and digits.  In this case, we use a mask attack assuming the following.

- Password length is 9 letters
- first letter is uppercase 
- next 4 digits are lowercase
- last four letters are digits

Ex. Abbbb1984

In [29]:
hc = hashlib.md5(b'Quant2016').hexdigest()
hc

'07cff3dbabc8fc112d4f8a3407b682b4'

In [30]:
r2 = '''
feed into hashcat - 
07cff3dbabc8fc112d4f8a3407b682b4acce:Quant2016

~42000 KH/s

takes 22 min

'''

### Using GPU and ASICs (Better Hardware)

Cheap nvidia GPU ($30) reaches MD5 speed of 400 MH/s

ASICs mades specifically for mining.  $200 used can get to 1,155 GH/s for mining

### Basics of RSA Encryption

https://en.wikipedia.org/wiki/RSA_(cryptosystem)
RSA is one of the first practical public-key cryptosystems and is widely used for secure data transmission. In such a cryptosystem, the encryption key is public and differs from the decryption key which is kept secret. In RSA, this asymmetry is based on the practical difficulty of factoring the product of two large prime numbers, the factoring problem. RSA is made of the initial letters of the surnames of Ron Rivest, Adi Shamir, and Leonard Adleman, who first publicly described the algorithm in 1977. Clifford Cocks, an English mathematician working for the UK intelligence agency GCHQ, had developed an equivalent system in 1973, but it was not declassified until 1997.[1]

### discussing private and public keys

Consider two small prime numbers

In [31]:
p = 61
q = 53

Compute Product

In [32]:
n = p * q
n

3233

Compute the TOTIENT of the product

In [33]:
t = (p - 1) * (q - 1)
t

3120

Choose any number 1 < e < t which is coprime to t


In [34]:
e = 17

Compute the modular multiplicative inverse d
d * e mod t = 1

In [35]:
d = 2753

check

In [36]:
d * e % t

1

#### Ready to look at keys

The PUBLIC KEY is n = 3233, e = 17, the ENCRYPTION FUNCTION for a m is:

c(m) = m^17 mod 3233

In [37]:
m = 77 # our message
c = m ** e % n # encryption
c

3123

The PRIVATE KEY is d = 2753. The DECRYPTION FUNCTION for a ciphertext message c is:

m(c) = c^2753 mod 3233

In [38]:
m = c ** d % n
m

77

### Working with REAL RSA key pairs

the library PyCryptoDome rovides a multitude of security algortihms and techniques for easy use with python

In [39]:
import sys
sys.path.append('c:\python\python35\lib\site-packages')
sys.path

['',
 'C:\\Users\\jonto\\Documents\\GitHub\\Python',
 'C:\\Python',
 'C:\\Users\\jonto\\Anaconda3\\python35.zip',
 'C:\\Users\\jonto\\Anaconda3\\DLLs',
 'C:\\Users\\jonto\\Anaconda3\\lib',
 'C:\\Users\\jonto\\Anaconda3',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\Sphinx-1.4.6-py3.5.egg',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\win32',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\win32\\lib',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\Pythonwin',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\setuptools-27.2.0-py3.5.egg',
 'C:\\Users\\jonto\\Anaconda3\\lib\\site-packages\\IPython\\extensions',
 'C:\\Users\\jonto\\.ipython',
 'c:\\python\\python35\\lib\\site-packages']

In [40]:
from Cryptodome.PublicKey import RSA

In [41]:
secret_code = 'PythonQuants'
key = RSA.generate(2048)

In [42]:
pub_key = key.publickey().exportKey()
print(pub_key)

b'-----BEGIN RSA PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAuJgaQyBAs4ozPfu6czUh\n8cH1oQVF3AADsyTa9yUbNrcrTI22vG/rkbyfvjFGnm5J68X1B0sTuFfaIDjPs9s4\nL9U0xOQql0qpUxfREofpaYVrn+3LTORTFQHlsr3hD20oS7gKdNCip9d+zaRFRQ/M\n9TiM7ZvdCsbXeEIW81QnnpFY352PeJIdkpOfl9DioM6EvKxvrFVosyOtmTv9zPsg\nXPxyGFk6j5PJmIIQOToEByMHPcYYRoyD3yI/UheWciGvWe8FvICZ0bS6vjbYnxnW\nHyNMLwQI4fLtB8ApWGddGlUzL/9903+YEGhkHlDkFOWcC3+f2rDSHDOI7T+4A0Yv\nnwIDAQAB\n-----END RSA PUBLIC KEY-----'


In [43]:
priv_key = key.exportKey()
print(priv_key)

b'-----BEGIN RSA PRIVATE KEY-----\nMIIEogIBAAKCAQEAuJgaQyBAs4ozPfu6czUh8cH1oQVF3AADsyTa9yUbNrcrTI22\nvG/rkbyfvjFGnm5J68X1B0sTuFfaIDjPs9s4L9U0xOQql0qpUxfREofpaYVrn+3L\nTORTFQHlsr3hD20oS7gKdNCip9d+zaRFRQ/M9TiM7ZvdCsbXeEIW81QnnpFY352P\neJIdkpOfl9DioM6EvKxvrFVosyOtmTv9zPsgXPxyGFk6j5PJmIIQOToEByMHPcYY\nRoyD3yI/UheWciGvWe8FvICZ0bS6vjbYnxnWHyNMLwQI4fLtB8ApWGddGlUzL/99\n03+YEGhkHlDkFOWcC3+f2rDSHDOI7T+4A0YvnwIDAQABAoIBACzkQK7GbzW6jE+s\nomFWMJUcuGGaaKziDARLGD02dvHNheguJpyZE07z8l1MmoH5DF0cXUSSy47vPorL\nhieVTorDbPvdCUaVz6v3hM7e4rLY0Z2pNOHUPShEt9nKN6uKvlv7u/9Ape3viFT2\neCodd5jDUfKPDyiJxujmGbK/aoszH971PA0/wWxxVfwkMytVFIt/QpZoUOF00kdl\nkzPRpOjtCcIG8kkTpG8p8jSadI41F0Z5M/iZRxVDsTN9jaNpByifCmNR5IvXJAyE\nmQrpZJybZ7vnCCQFlV2iyZLDRL2NDY1ysf52nVDnYXlQDAIyvy+FImXhSJf4Mzni\nsuctiRECgYEAzKrX/P/jajAwYhoQIUDuOt3SitgIq6xQlNKViDw3fZojrxhKWA36\nhthrlrOp0MB3OWU/o/ay2fEq/cLJPoi1Ovacda6minJI8zs6qIDQmgpilB3qke9V\noOMv0CB5USqapD40R286uFcOb/UXJHy9DRDF8xeKEMeKDUe6KfXxo2UCgYEA5uRp\nIkC5hA/Qxx7zXMze2E0bAkK6rgc3Ahez4K2haVhzh

Let us now encrypt and decrypt  message with these keys.  First, the encryption step using the public key - assuming that someone sends a message to the Python Quants who own the private key

In [44]:
from Cryptodome.Cipher import PKCS1_OAEP

In [45]:
m = b'Hello to the Python Quants,  ' * 5

In [46]:
cipher = PKCS1_OAEP.new(key.publickey())   # public key

In [47]:
ciphertext = cipher.encrypt(m)
ciphertext

b'\x0e\xb2\xcb\x03R}p\xe2\x1c\x95\xcc\x1b\x93\xe5k8\x8d\xc1UF\xab\x9e$OI\xff\xcaXk\xcc\xe0)\xda\xbb\xbe\xab\x93\x124\xfbpL\x1b\xf1p\xff\xe5\xd7\xcd\x7f/\x07Mh\xdd\xd9\x11nZy5\xef\x1d\xfe\x9e\xd3\xef\xa0\xee,B\x06\xaa\xe4^\xd6\xe6\xd14U\xb2\x94\xa5Q\x95\x94P\xda\xb8\x93\xd4\xb7Q\xe2\x972D5\x8a\x16\xcfT\xd9:oW\x88\x8c\xa1\x19\xe4\x9f\xc6r\xd5\xe0\x98Z8\x1e\xa8\xb1\xc0\x08\xd6\xd3\x1cM\x92\x11\xe9d\xe9\xc0\x92c\x15Lz\x9cA<\x7f\xd4^!H\xbf\xc6/\x80\xc3\x8ab\x15gs\x9b\x87\x99\xaf\xe1{3\xea1:3r\x91q\x86\\\x8e\xca\xe7\xdd\xefn\x8c\x08\xab<\rj^\xe8\x17h\xda?X\xc7\x8d\x08#\x99\xc1\xb6\x0ez\xaeCe\x01r\x14\x12SM\xec\xb44\xf6\xa7c"-HW\xf0\xb6\x00\xe5\xf3\xa0\xaa\x8d\x9e\xa2T||\x01"\xe1\x97\x1a\x9d\xdf0\x14{2\x9b\x81\x83\xac\xed\xe8\'\xd6w@\xb97'

Second, the decryption step using the private key

In [48]:
cipher = PKCS1_OAEP.new(key)   # private key

In [49]:
message = cipher.decrypt(ciphertext)
message

b'Hello to the Python Quants,  Hello to the Python Quants,  Hello to the Python Quants,  Hello to the Python Quants,  Hello to the Python Quants,  '

### Using openssl

Beyond Python, it is easy to generate RSA key pairs by the use of openssl

In [50]:
# !openssl rq -x509 -nodes -days -365 -newkey rsa:1024 -keyout cert.key -out cert. pem -b

## Signing messages with hashes and RSA keys

Let us electronically sign a message. The Python Qyuants send a essage to someone else.  To this end, we combine hashing with RSA encryption

In [51]:
m = b'Hello FROM the Python Quants. ' * 5

We generate a HASH CODE for the message

In [52]:
from Cryptodome.Hash import SHA256

In [53]:
h = SHA256.new(m)

In [54]:
hd = h.hexdigest()
hd

'f226dcc2ff1b6ae0e233b0f5b5c8ebe5c32ed7fdbcb078f50fd81f568efc16e0'

We next SIGN THE MESSAGE i.e. we encrypt the hash code with the Private Key from before

In [55]:
from Cryptodome.Signature import PKCS1_v1_5

In [56]:
signer = PKCS1_v1_5.new(key)

In [57]:
signature = signer.sign(h)  # private key
signature

b'\\\xfdQ\x8f8\xceO\xdb*\x8d\xc0\x84\xfa\xe2\xed\xfb\x9c\x97;\x01\x14\r\xc8db+\x15\xd8O\xe9M\xab\xc02!\x1cJ/\xbd\xc72\xcb}\xb6\x8c\xf0 D[\x03Sg\x01\xcaQm\x81\x1cj\xec(\xd7\xc1\x00\xf99D\xf1i\x87\xec\x86(\xb4\x94\xd5W\xd1\x8cJ-_\xa5J\xa2\xe4\x92*\x08\xa9q\x0f>\\#\xad\xe4\xeb\xda\xfc\n\xfb\xc7`0\xcb\x89\xd2\xfd\xefZ\xe8U\xddbk\xbf5s\x0e\xd7\xb2}fPX\x0e\x0f=\x1b(O\x10l:\xf5t\x8cg\x832\x0f{ \xc0\x84\xd9R\xe8\'*\xa0\x07\x9c\xf8\x87\x89g\xa0c\x14Rx\xd9\\\x8b*\x05\xa2\x17\xac|+\xa5k\x9c3\xfc\xb2\xc5A\x15\xa7=\x05OA"?{(\x91\x85, \x03\xdd\xbaJ\xde\xf8\xd0\x84["9\x83\xfe\xe2\x14\xc7X\xfc\xd2w\\W\x8d\xd5\xf9\x19\xce)p\xc8T\t+\x9a=\xb21et\xa3\xe4^\xb0\xd89\xeeX\xf6KA"\xa7\xf7w\x9f\x98X\xbbMv6'

Someone else, the receiver of the message, who knows our public key can now verify that we ahve signed the message as follows

In [58]:
hashcode = SHA256.new(m)
hashcode.hexdigest()

'f226dcc2ff1b6ae0e233b0f5b5c8ebe5c32ed7fdbcb078f50fd81f568efc16e0'

In [59]:
PKCS1_v1_5.new(key.publickey()).verify(hashcode, signature)  # public key

True

## Basic idea behind blockchains

#### Very simple example

Lets now illustrate a blockchain using an example - Recall the properies ofhash functions that we require (collision resistance, hiding, puzzle friendliness).  Lets focus on COLLLISION RESISTANCE and HIDING for the moment

As a starter, its easy to calculate the hash value of a string ("first block"), for instance....

In [60]:
import hashlib

In [61]:
b1 = 'Mike 1981'      # Mike kid #1
h1 = hashlib.md5(b1.encode('ascii')).hexdigest()   # hash for first block
h1

'2adccd590d7c3f8a862df599ae8eac49'

It is highly unlikely that another input yields the same output.  It is also really difficult (Ner Impossible) to find the input given a certain output

A block chain can be used to document events over time (transactions, kids).  To this end, we take the hash code from the first block, add the second block information and calculate a new hash value:

In [62]:
b2 = h1 + ', Jon 1984'
h2 = hashlib.md5(b2.encode('ascii')).hexdigest()
h2

'7093f419d367de9d1d7715cf001bcfdb'

A third block is easily added

In [63]:
b3 = h2 + ', Zack 1991'
h3 = hashlib.md5(b3.encode('ascii')).hexdigest()
h3

'ea93e313e0458f2cb346e48fe6eb8f9f'

Our block chain is now:

In [64]:
print(b1)

Mike 1981


In [65]:
print(b2)

2adccd590d7c3f8a862df599ae8eac49, Jon 1984


In [66]:
print(b3)

7093f419d367de9d1d7715cf001bcfdb, Zack 1991


There is one MAJOR PROBLEM - the chain is easy to manipulate since you only need to re-calculate the whole chain.  You have all the information needed

We need to add a security measure to avoid manipulation: SIGING OF THE LAST HASH VALUE

In [67]:
h3

'ea93e313e0458f2cb346e48fe6eb8f9f'

In [68]:
# key generation
from Cryptodome.PublicKey import RSA
key = RSA.generate(2048)

In [69]:
# signing
from Cryptodome.Hash import MD5
from Cryptodome.Signature import PKCS1_v1_5
md5 = MD5.new(b3.encode('ascii'))
signature = PKCS1_v1_5.new(key).sign(md5)    # private key
signature

b'\x11\x8blZ\xf2\x07\xef\xe9Xv\xf4\xce\\\x12\xd6\xfd3$7\x88\xba\x8f<\xb8\x07l8\xa48\x9e\xc6\xec\xbb"\xd9\x92\xa5N\xdc\xc5vZ\xa53\n\x06\xedq\x1b\x0e\xbc\x92q\xc7\xff\x01\x8d\xc8\xe6t,\x9b\x06\xe9lv\xc4\x9d\xcc\xba\x8a\xc2\xbce\x89\xee\xcef\xbd!C\x9e}y\xe1\xf9\xfa\x05\x8ca%\x19\xea\xe4\xfdWd\x01\xd2\xb9.\xb4\x98\x00\xd9\xb5D\x8c\xa3\xfa\x92\x13\x90\xdf\xf7\x9d\xfcn\xd2!\xbd\x83\x8c\x17w\xc9\x8a\x8c\xf9@\x7f\xaa\xa8f\xef\x9c0\x112;\x99\xa7\\Z\xfd[\xcd\x19\x035\x98/\x9e\x92[\xf7\x0e\xb6\xf4qurc]\x13Rk\x01S\x90G\t\xbd\xea/\xc6\xe0\x1f{Yl0\xf18\xf9\x8e\x9e\x04\xc2[\x1et\xfcry\x1b\xe3\x9b[e\x1f&h\x87c\xa9\xf2\xe3\x9dpu\xa4\xa3\xa1r\xd8\x8e\x16d\xe7m\xe2H\xf2\xfb)\x05\x9b\x15\xcc\xbbi\xc7\x18\xc5,\x19\x0e_\x1ae\xcb\x04\xe7\x17#\xd7)d~\x8f\x12\xc26\x87\xb1'

If we can make sure that the private key is safe, then the block chain plus the signature for the final hash value are ALMOST IMPOSSIBLE to manipulate - although all the infromation is publicly available

In [70]:
b1, b2, b3, h3, signature

('Mike 1981',
 '2adccd590d7c3f8a862df599ae8eac49, Jon 1984',
 '7093f419d367de9d1d7715cf001bcfdb, Zack 1991',
 'ea93e313e0458f2cb346e48fe6eb8f9f',
 b'\x11\x8blZ\xf2\x07\xef\xe9Xv\xf4\xce\\\x12\xd6\xfd3$7\x88\xba\x8f<\xb8\x07l8\xa48\x9e\xc6\xec\xbb"\xd9\x92\xa5N\xdc\xc5vZ\xa53\n\x06\xedq\x1b\x0e\xbc\x92q\xc7\xff\x01\x8d\xc8\xe6t,\x9b\x06\xe9lv\xc4\x9d\xcc\xba\x8a\xc2\xbce\x89\xee\xcef\xbd!C\x9e}y\xe1\xf9\xfa\x05\x8ca%\x19\xea\xe4\xfdWd\x01\xd2\xb9.\xb4\x98\x00\xd9\xb5D\x8c\xa3\xfa\x92\x13\x90\xdf\xf7\x9d\xfcn\xd2!\xbd\x83\x8c\x17w\xc9\x8a\x8c\xf9@\x7f\xaa\xa8f\xef\x9c0\x112;\x99\xa7\\Z\xfd[\xcd\x19\x035\x98/\x9e\x92[\xf7\x0e\xb6\xf4qurc]\x13Rk\x01S\x90G\t\xbd\xea/\xc6\xe0\x1f{Yl0\xf18\xf9\x8e\x9e\x04\xc2[\x1et\xfcry\x1b\xe3\x9b[e\x1f&h\x87c\xa9\xf2\xe3\x9dpu\xa4\xa3\xa1r\xd8\x8e\x16d\xe7m\xe2H\xf2\xfb)\x05\x9b\x15\xcc\xbbi\xc7\x18\xc5,\x19\x0e_\x1ae\xcb\x04\xe7\x17#\xd7)d~\x8f\x12\xc26\x87\xb1')

Antoher idea is to make it hard ot construct another block chain with the same inputs or other inputs that meet a certain requirement.  Let us define that only hash values with FIVE ZEROS at the end are allowed.  To this end, we must allow for an additional input parameter

In [71]:
%%time
n = 0
while True:
    b = str(n) + ', ' + b1
    md5 = hashlib.md5(b.encode('ascii')).hexdigest()
    if md5[-5:] == '00000':
        print(b + ' --> ' + md5)
        break
    n +=1

2236417, Mike 1981 --> f60eb6c18f5a01c9e60a1f6319000000
Wall time: 4.73 s


Someone wnting to manipulate the block chain must put in much more effort in this case than without such a requirement.  The difficulty can easily be increased by requiring more trailing zeros

In [72]:
%%time
n = 0
while True:
    b = str(n) + ', ' + b1
    md5 = hashlib.md5(b.encode('ascii')).hexdigest()
    if md5[-6:] == '000000':
        print(b + ' --> ' + md5)
        break
    n +=1

2236417, Mike 1981 --> f60eb6c18f5a01c9e60a1f6319000000
Wall time: 4.32 s


The first secuity measure SIGNING is vulnerable to STEALING the private key.  The second one TARGETING to sheer BRUTE FORCE.  A combination of both works well and is more secure.

Another approach that adds some security is to use RANDOM, PUBLICLY KNOWN, FIXED INITIAL HASH

In [73]:
import os
h0 = os.urandom(16).hex()
h0

'38757edf8e81382d1d2fed3e850965a1'

In [74]:
b1 = h0 + ', Mike 1982'
h1 = hashlib.md5(b.encode('ascii')).hexdigest()
h1

'f60eb6c18f5a01c9e60a1f6319000000'

## Simple Cryptocurrency based on a block chain

Let us now implement a simple cryptocurrency based on the blockchain idea.  The example is from the book Narayanan et al: Bicoin and Cryptocurrency Technologies

Consider a central authority, The Python Quants (TPQ), that issues a cryptocurrency called TPQ COINS.  Two basic transactions are possible CreateCoins and PayCoins in the system.

There are many other paticipants in the system that are identified by their public key of an RSA key pair.

Consider an original ransacion by which PQ creae 10 TPQ Coins.  We have transaction id of 0.  The coins created have a coind id of 0(0).  The value is 10 and the recipient are TPQ represented by their public key.  we also generate an initial hash value to be included in the first block.  We define the initial block as follows

In [75]:
# h = h0 = os.urandom(16).hex()
h = hashlib.md5('my secret passphrase'.encode('ascii')).hexdigest()
# prev, hash, ransID, transaction, value, coinID, recipient
b0 = h + '\n0, CreateCoins, 10, 0m, 0xTPQ'
print(b0)

f83e9190098974434be56e5738d89f5f
0, CreateCoins, 10, 0m, 0xTPQ


this initial block gets hashed and he hash alue gets signed by PQ

TPQ now pays another particiapnt 5 TPQ Coins.  This is PayCoins transaction with id 1

In [83]:
h = hashlib.md5(b0.encode('ascii')).hexdigest()
b1 = h + '\n1, Consumed Coins: 0a\n'
# prev hash, tansid, transaction, value, transid, recipient
b1 += '1, PayCoins, 5, 1a, 0xRE1\n'  
b1 += '1, PayCoins, 5, 1b, 0xTP2'    # another address of TPQ
print(b1)

0f57fe8e108c77623333774e56066544
1, Consumed Coins: 0a
1, PayCoins, 5, 1a, 0xRE1
1, PayCoins, 5, 1b, 0xTP2


The second block gets hashed and the hash value gets assigned by TPQ

Now, the participant with the address (public key) of 0xRE1 is able to spend TPQ coins to buy good sor services from someone identified as 0xSEL

In [84]:
h = hashlib.md5(b0.encode('ascii')).hexdigest()
b2 = h + '\n2, Consumed Coins: 1a\n'
# prev hash, tansid, transaction, value, transid, recipient
b2 += '2, PayCoins, 2.5, 2a, 0xSEL\n'  # payment to seller
b2 += '2, PayCoins, 2.5, 2b, 0xRE2'    # another address of 0xRE1
print(b2)

0f57fe8e108c77623333774e56066544
2, Consumed Coins: 1a
2, PayCoins, 2.5, 2a, 0xSEL
2, PayCoins, 2.5, 2b, 0xRE2


This block gets hashed and signed by the payer 0xRE1.  It then gets added to the blockchain by TPQ, signed by TPQ, and published.  Everybody can now recover the complete history of the TPQ coins.

the blockchain for TPQ now looks as follows

In [85]:
print(b0)

f83e9190098974434be56e5738d89f5f
0, CreateCoins, 10, 0m, 0xTPQ


In [86]:
print(b1)

0f57fe8e108c77623333774e56066544
1, Consumed Coins: 0a
1, PayCoins, 5, 1a, 0xRE1
1, PayCoins, 5, 1b, 0xTP2


In [87]:
print(b2)

0f57fe8e108c77623333774e56066544
2, Consumed Coins: 1a
2, PayCoins, 2.5, 2a, 0xSEL
2, PayCoins, 2.5, 2b, 0xRE2


## Bitcoin Addresses

In [90]:
import os

import ecdsa

import hashlib

import base58

import binascii





class Key(object):



    def __init__(self, private_key=0):

        if private_key == 0:

            self.private_key = os.urandom(32)

            self.printable_pk = str(binascii.hexlify(self.private_key), "ascii")

        else:

            self.printable_pk = private_key

            self.private_key = binascii.unhexlify(private_key.encode('ascii'))





        self.sk = ecdsa.SigningKey.from_string(self.private_key, curve = ecdsa.SECP256k1)

        self.vk = self.sk.verifying_key

        self.public_key =  b"04" + binascii.hexlify(self.vk.to_string())

        ripemd160 = hashlib.new('ripemd160')

        ripemd160.update(hashlib.sha256(binascii.unhexlify(self.public_key)).digest())

        self.hashed_public_key = b"00" + binascii.hexlify(ripemd160.digest())

        self.checksum = binascii.hexlify(hashlib.sha256(hashlib.sha256(binascii.unhexlify(self.hashed_public_key)).digest()).digest()[:4])

        self.binary_addr = binascii.unhexlify(self.hashed_public_key + self.checksum)

        self.addr = base58.b58encode(self.binary_addr)



    def get_addr(self):

        return self.addr





def get_private_key(private_key):

    return private_key



def generate_new_private_key():

    return os.urandom(32)



def generate_sk(private_key):

    return ecdsa.SigningKey.from_string(private_key, curve=ecdsa.SECP256k1)



def generate_vk(sk):

    return sk.verifying_key



def generate_public_key(vk):

    return b"04" + binascii.hexlify(vk.to_string())



def generate_hashed_public_key(public_key):

    ripemd160 = hashlib.new('ripemd160')

    ripemd160.update(hashlib.sha256(binascii.unhexlify(public_key)).digest())

    return ripemd160.digest()





def generate_hashed_public_key_string(public_key):

    ripemd160 = hashlib.new('ripemd160')

    ripemd160.update(hashlib.sha256(binascii.unhexlify(public_key)).digest())

    return binascii.hexlify(ripemd160.digest()).decode('ascii')

    #return ripemd160.digest()





def generate_checksum(hashed_public_key):

    return binascii.hexlify(hashlib.sha256(hashlib.sha256(binascii.unhexlify(hashed_public_key)).digest()).digest()[:4])



def generate_address(private_key):

    sk = generate_sk(binascii.unhexlify(private_key.encode('ascii')))

    vk = generate_vk(sk)

    public_key = generate_public_key(vk)

    hashed_public_key = generate_hashed_public_key(public_key)

    checksum = generate_checksum(hashed_public_key)

    binary_address = binascii.unhexlify(hashed_public_key + checksum)

    addr = base58.b58encode(binary_address)

    return addr

An address can be generated based on a random private key

In [91]:
key = Key()

In [92]:
key.printable_pk

'b6880f108ded0a5d6dded413be34e97bb20615c057af6aebcf8fc995222c1b5c'

Having generated the private key, the public key is generated then via ECDSA, i.e. via multiplying the private key with the respective generator point G (x,y) of they elliptic curve.  The idea is that the multiplication is easy and straightforward when the factorization is an almost impossible to solve problem.

In [93]:
key.public_key

b'042edd33b371f32027ea1b10e47a51762e5834ffec4d8ccb43fae190d14f95309a1fb70d2f658c95c5a8c683157b45ab08710ceb6afbc613e7134726391b6ddc98'

Bitcoin addresses are represented int he Wallet Import Format (WIF) which is based on the base58Check encoding format

In [94]:
key.addr

'16ba6cMMNupsuU3QQupbMZddzsjWVQbFgg'

Of course you can create your own private key based on a passphrase

In [98]:
pk = hashlib.sha256('Bitcoin is really cool.'.encode('ascii')).hexdigest()
pk

'3098b5f8ae81c805cc4ab9e43c3ae08e37a2e0a2901abb9804eb55613a60d25b'

In [99]:
key = Key(pk)

In [100]:
key.printable_pk

'3098b5f8ae81c805cc4ab9e43c3ae08e37a2e0a2901abb9804eb55613a60d25b'

In [101]:
key.public_key

b'045bbb1c8c7ca479c705885d8a40ac14899a8bff51437d59eb2229fc1dfee86843267ec1788740d11783360813322c28c2ac78080e30bbca747b8b95c3010f4aa2'

In [102]:
key.addr

'19QiM7YmzSV6fARZnVvaV5u9Y8gqw56g25'

## Bitcoin Transactions

## Basics behind mining

Mining is based on SHA256 hash codes

In [104]:
sha256 = hashlib.sha256('jonto'.encode('ascii'))
sha256.hexdigest()

'465bde7c670b2cc20c2c65c63c8f59027f44088ce753663894b755acad1726ee'

the idea behind mining is to find a hash code small enough, i.e. lies below a certain target level (defined by leading zeros)

In [105]:
target = '%064x' % (1000000000 << 200)
target

'0000003b9aca0000000000000000000000000000000000000000000000000000'

the original hash for python is not small enough

In [106]:
target

'0000003b9aca0000000000000000000000000000000000000000000000000000'

In [107]:
sha256.hexdigest() < target

False

However, adding certain numbers to the string yields a hashcode small enough

In [109]:
sh = hashlib.sha256(b'%dyves' % 23240167).hexdigest()
sh

'00000003b04fad4b30a527760fea6ee5beec8035ef636316c2bf2577b2789611'

In [111]:
sh < target

True

The following simulates mining

In [113]:
%%time
i = 0
while True:
    sha256 = hashlib.sha256(b'%d' % i + b'yves')
    if sha256.hexdigest() < target:
        print('SUCCESS')
        print(i, sha256.hexdigest())
        # break
    if i % 2500000 == 0:
        print(i)
    i += 1
    if i > 55000000:
        break

0
2500000
5000000
7500000
10000000
12500000
15000000
17500000
20000000
22500000
SUCCESS
23240167 00000003b04fad4b30a527760fea6ee5beec8035ef636316c2bf2577b2789611
25000000
SUCCESS
27090678 00000007a53f0b5163e1cb7ade64881e4eb3e06f9c102ad72a19c942223ba82b
27500000
30000000
32500000
35000000
37500000
40000000
42500000
45000000
47500000
50000000
SUCCESS
50427211 000000099f6760417f3b2161a9ba9e989c62da1745d949267d5b648edaa21496
52500000
55000000
Wall time: 1min 17s


## Mining a Bitcoin Block

In [114]:
import struct
import binascii
import hashlib

check back to video for rest